### ISSUE TITLE
Modernize auth/key handling

### ISSUE THREAD

**Issue by bitprophet** [fabric/fabric#1618]
## Immediate problem

Fab v2 currently stinks for real-world auth situations where you have multiple keys (say, github + server, or personal + work, or combinations thereof, or etc), and/or multiple sources for those keys (such as `~/.ssh`, an SSH agent, or in-memory keys taken out of a secrets store, etc). This is because none of the auth handling from v1 has been ported. Thus, it's quite easy to encounter Paramiko-level `AuthenticationException` or `SSHException`. (The best method to avoid that, at present, is to rely on how Paramiko tries all agent keys, as all key types, before giving up -- and thus to make sure the right key is early in your agent list.)

The deeper problem is that the auth handling from v1 is pretty gross, because Paramiko's auth handling is somewhat gross, and it'd be _nice_ not to just port it directly.

## Background

There are a handful of related issues (arguably some should already be closed as dupes) in the [Auth/Keys label](https://github.com/fabric/fabric/labels/Authentication%2FKeys) that have been warts forever. High level/consolidated overview:

- First, Paramiko's own auth handling is shit, and needs improvements. (See paramiko/paramiko#387.)
  - Fabric 1 has to guess constantly about what a given Paramiko exception instance truly means, based on what parameters the Fabric level has handy (such as presence or lack of a password value).
  - We can't usefully improve until Paramiko raises usefully distinct errors, at the very least, such as an aggregate error noting which keys were tried and how each of them failed (e.g. "I tried <some RSA in-memory pkey> and it was rejected; and I tried <some EC from-file pkey> but it appeared to be encrypted and you didn't give a passphrase; finally I tried <some agent key> but it didn't actually seem to be ANY key type I recognize")
  - Possibly also want to clean up the params of `Client.connect` (though those are not super duper awful, on their own.)
    - Though at least one of them *is* immediately problematic: `password`, which is used both for actual plaintext password auth *and* for key decryption. This needs to change in order for us to limit the number of misleading exceptions.
- Aside from that, Fabric v1 has its own old, naive assumptions about the world:
  - That there's only ever a single explicit username, password and/or key to use globally instead of on a per-host basis.
    - This ties strongly into #4, and is already mostly-solved simply by having `Connection` objects which carry their own non-shared config data;
    - Though the CLI-flag use case still has issues, since there's no good way to 'tag' flags as being 'for' a given host. That said, config files (SSH and Invoke/Fabric level) are a thing, so...
  - The `password` setting does double duty as both true *password* auth as well as key *passphrase*. This has always been gross. EDIT: actually, this is Paramiko's fault too! It's where that behavior is implemented.
    - At the very least, these concepts need to be made into distinct CLI flags and config settings;
    - Presumably including multiple "prompt me up front" options, as we had already in v1 for the one password setting.
      - And going by my thoughts around Invoke's implementation of `sudo`, we may well want to do away with 'lazy' runtime prompting entirely.
- Then there's other random stuff that probably needs to be their own tickets:
  - Support for an OS-level keychain for key passphrases (i.e. OpenSSH `UseKeychain`, at least on OS X Sierra) - this likely wants to be in Paramiko like agent support.

## Solution brainstorm

- Ideal: get out a Paramiko 3.0 with an overhauled auth handler logic core, and wrap it with a nice, gentle, compact bit of logic at the Fabric level. Sadly this will take time.
- Fast but crappy: just port v1's `connect` logic which loops-until-success and does chicken-entrail level divination about any resulting Paramiko errors.
- Meet halfway: implement enough functionality (which is, typically, required either way) to let users specify very exact auth parameters (including up-front passphrase entry and/or keychain integration) so they can ensure Paramiko gets the right key or password for a given host.

**Comment by radssh** [comment]
Hi @bitprophet - hopping in from the Paramiko side 😃

Not sure where you want to officially track the underlying overhaul of the "terribad" (love that term, btw) key/auth handling that's been lingering, and want to volunteer to assist in addressing a potential cleanup/reorg/rearchitect to stop the accumulation of tacking on the "good enough for now" cumulative bits and pieces. Probably would take a Paramiko 3.0 goal to break loose from the chains of maintaining backward compatibility - I've tinkered with some ideas, but that compatibility thing, yuck...

This topic seems better suited to the Paramiko issue, so I chose to drop this here, but can continue the dialog wherever you deem most appropriate.

**Comment by bitprophet** [comment]
Update on this:

- Last month I solved the largest practical issue here, namely Fabric 2's inability to affect much of Paramiko's auth settings. All Paramiko-level connect kwargs can be arbitrarily configured via the Invoke config system; SSH config is now honored for a bunch of settings; new CLI args were added for configuring identity or submitting up-front interactive secrets; etc.
- Part of that is requiring Paramiko 2.4, which added a `passphrase` kwarg to help move away from the abovementioned conflation problems in `password`.
- I'm approaching this from a "partway-through-the-session prompting is an antipattern" perspective. Right now, v2 acts like v1 with `--abort-on-prompts` turned on: if you didn't prep the connection with all the info it needs, it'll raise an exception instead of prompting.
  - Ironically, this means paramiko/paramiko#387 _isn't strictly necessary_ for v2, since the top reason to work on it was to remove Fab 1's guessing-game around how/whether to prompt the user on an auth exception.
  - However, I'm still going to wrap that up before v2 launches - not only is it (non-crap exception behavior) what I'd personally expect from a v2, but I'd like to leave the door open re: adding "prompt me partway" back in sometime after 2.0 if enough folks say they'd find it useful.
